lib/babe/epoch.go

```diff
@@ -221,12 +227,35 @@ func (b *Service) incrementEpoch() (uint64, error) {
 // It returns the wrapped error errOverPrimarySlotThreshold
 // if it is not authorised.
 // output = return[0:32]; proof = return[32:96]
-func (b *Service) runLottery(slot, epoch uint64, epochData *epochData) (*VrfOutputAndProof, error) {
-	return claimPrimarySlot(
+func (b *Service) runLottery(slot, epoch uint64, epochData *epochData) (*types.PreRuntimeDigest, error) {
+	proof, err := claimPrimarySlot(
 		epochData.randomness,
 		slot,
 		epoch,
 		epochData.threshold,
 		b.keypair,
 	)
+	if err == nil {
+		preRuntimeDigest := types.NewBabePrimaryPreDigest(epochData.authorityIndex, slot, proof.output, proof.proof)
+
+		return preRuntimeDigest.ToPreRuntimeDigest()
+	}
+
+	if errors.Is(err, errOverPrimarySlotThreshold) {
+		if epochData.secondary == 0 {
+			return nil, errSecondarySlotProductionDisabled
+		}
+
+		proof, err = claimSecondarySlot(epochData.randomness, slot, epoch,
+			epochData.authorities, b.keypair, epochData.authorityIndex)
+		if err != nil {
+			return nil, fmt.Errorf("cannot claim secondary slot: %w", err)
+		}
+		if proof != nil {
+			preRuntimeDigest := types.NewBabeSecondaryPlainPreDigest(epochData.authorityIndex, slot)
+			return preRuntimeDigest.ToPreRuntimeDigest()
+		}
```

Comment:
proof isn't used for secondary plain, only secondary VRF, so you don't need to return the proof from `claimSecondarySlot`, maybe you can update it to be `claimSecondarySlotPlain` and not return any proof?

Reply:
done